src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java

```diff
@@ -47,51 +76,75 @@ public int getChunkZ() {
         return cz;
     }
 
+    private boolean isInvalid(int x, int y, int z) {
+        if (y < 0 || y >= worldHeight) return true;
+        return ((x | z) & ~0xF) != 0;
+    }
+
     @Override
     public boolean isTrue(int x, int y, int z) {
-        return store[x][z][y];
+        if (isInvalid(x, y, z)) return false;
+        int status = data.get((y * PAGE_SIZE) + (x >> 1));
+        return ((status >>> (z + (((x & 1) == 1) ? 16 : 0))) & 1) != 0;
     }
 
-    @Override
-    public void setTrue(int x, int y, int z) {
-        if (y >= store[0][0].length || y < 0)
+    private void set(int x, int y, int z, boolean val) {
+        if (isInvalid(x, y, z))
             return;
-        store[x][z][y] = true;
+        int point = (y * PAGE_SIZE) + (x >> 1);
+        int bit = z + (((x & 1) == 1) ? 16 : 0);
+        int source;
+        do {
+            source = data.get(point);
+            int target = source;
+            if (val) {
+                target |= 1 << bit;
+            } else {
+                target &= ~(1 << bit);
+            }
+            if (data.compareAndSet(point, source, target)) break;
+        } while (true);
         dirty = true;
     }
 
+    @Override
+    public void setTrue(int x, int y, int z) {
+        set(x, y, z, true);
+    }
+
     @Override
     public void setFalse(int x, int y, int z) {
-        if (y >= store[0][0].length || y < 0)
-            return;
-        store[x][z][y] = false;
-        dirty = true;
+        set(x, y, z, false);
     }
 
     @Override
     public boolean isEmpty() {
-        for (int x = 0; x < 16; x++) {
-            for (int z = 0; z < 16; z++) {
-                for (int y = 0; y < store[0][0].length; y++) {
-                    if (store[x][z][y]) {
-                        return false;
-                    }
-                }
-            }
+        int end = data.length();
+        while (end-- > 0) {
+            if (data.get(end) != 0) return false;
         }
         return true;
     }
 
     @Override
     public void copyFrom(ChunkletStore otherStore) {
-        for (int x = 0; x < 16; x++) {
-            for (int z = 0; z < 16; z++) {
-                for (int y = 0; y < store[0][0].length; y++) {
-                    store[x][z][y] = otherStore.isTrue(x, y, z);
+        if (otherStore instanceof PrimitiveChunkStore) {
+            AtomicIntegerArray data = ((PrimitiveChunkStore) otherStore).data;
+            AtomicIntegerArray to = this.data = new AtomicIntegerArray(data.length());
+            int end = to.length();
+            while (end-- > 0) {
+                to.set(end, data.get(end));
+            }
+        } else {
+            int wh = worldHeight;
+            for (int x = 0; x < 16; x++) {
+                for (int z = 0; z < 16; z++) {
+                    for (int y = 0; y < wh; y++) {
+                        set(x, y, z, otherStore.isTrue(x, y, z));
+                    }
```

Comment:
Please explain the operations you are taking here, especially the thoughts behind the nested for loops and the while loop.